content/blog/2024/11/how-to-create-an-esp-idf-component/index.md

```diff
@@ -0,0 +1,494 @@
+        if(err != ESP_OK) {
+            ESP_LOGE(TAG, "Failed to read data from SHTC3 sensor");
+        } else {
+            ESP_LOGI(TAG, "Temperature: %.2f C, Humidity: %.2f %%", temperature, humidity);
+        }
+        vTaskDelay(1000 / portTICK_PERIOD_MS);
+    }
+}
+```
+
+On this task, the sensor will be woken up, the readout will be read and processed, and then the sensor will return to sleep mode.
+
+### Creating Kconfig (optional)
+
+As a bonus, this component has a dedicated configuration menu using a `Kconfig` file.
+
+Kconfig is a configuration system used in software projects to define and manage configuration options in a structured and hierarchical way. It's widely used in projects like the Linux kernel, Zephyr, and ESP-IDF (Espressif IoT Development Framework).
+
+On this configuration menu, you will be able to set the I2C GPIOs (SDA and SCL), the I2C port number, and the bus frequency. This removes any hardcoded configuration from the source code.
+
+```text
+menu "Driver SHTC3 Sensor"
+            
+    menu "I2C"
+        config SHTC3_I2C_NUM
+            int "I2C peripheral index"
+            default -1
+            range -1 3
+            help
+                For auto select I2C peripheral, set to -1.
+
+        config SHTC3_I2C_SDA
+            int "I2C SDA pin"
+            default 17
+            range 0 55
+            help
+                Set the I2C SDA pin for the data signal.
+
+        config SHTC3_I2C_SCL
+            int "I2C SCL pin"
+            default 18
+            range 0 55
+            help
+                Set the I2C SCL pin for the clock signal.
+
+        config SHTC3_I2C_CLK_SPEED_HZ
+            int "I2C clock speed (Hz)"
+            default 100000
+            range 10000 400000
+            help
+                Set the I2C clock speed in Hz.
+    endmenu
+  
+endmenu
+```
+
+The `Kconfig` file should be placed on the component root directory. To set the values, you can use the command:
+
+```bash
+idf.py menuconfig
+```
+
+To configure the I2C, go to the `Component config` -> `Driver SHTC3 Sensor` and set the `I2C SDA pin`, `I2C SCL pin`, and the `I2C clock speed (Hz)`.
+
+## Running the application
+
+To run the application, run the command to `flash` and `monitor` using the `idf.py`. The application will print the temperature and humidity.
+
+```bash
+idf.py flash monitor
+```
+
+### Console log output
+
+{{< asciinema
+  key="component_shtc3"
+  idleTimeLimit="2"
+  speed="1.5"
+  poster="npt:0:09"
+>}}
+
+## Conclusion
+
+Creating a new component requires several steps and might be confusing. In this article, we guided you on the basic steps of creating the component, adding a new component to this project, adding the code to the component, and finally adding the configuration menu.
+
+Writing a new component is highly recommended for improving the project architecture. This is a very powerful approach and can be used for many different types of component, not only for drivers.
+

```

Comment:
My previous comment was about mentioning some facts, like the sensor name. Now I think it is not needed. Still the conclusion can be slightly more refined like this:

```suggestion
In this article, we went through the basic steps of creating a component skeleton, adding it to a project, writing the code, testing it, and, finally, adding the configuration menu.

Even though creating a new component requires a few additional steps, it helps improve the project architecture and make code reusable. This is a very powerful approach not only for creating driver components, but any other code that can work in a modular fashion.
```